Getting started with Charmed OSM (#1088)
canonical-web-and-design/tutorials.ubuntu.com: tutorials/telco/charmed-osm-get-started/charmed-osm-get-started.md

---
id: charmed-osm-get-started

> summary: Getting started with Charmed OSM

Reviewer: ```suggestion
summary: An introduction to Open Source MANO deployments and VNF workloads orchestration with Charmed OSM. 
```

categories: NFV
tags: OSM, MANO, NFV, MicroK8s, MicroStack, beginner
difficulty: 1
status: draft
published: 2019-11-09
author: Tytus Kurek <tytus.kurek@canonical.com>
feedback_url: https://bugs.launchpad.net/canonical-osm

---

# Getting started with Charmed OSM

## Introduction

Duration: 5:00

This tutorial provides an introduction to Charmed OSM. It uses Canonical's Development Stack for [OSM][osm_website] (Open Source MANO (Management and Orchestration)) which is a toolset that you can use to onboard and orchestrate real VNF (Virtual Network Function) workloads directly on your workstation. It consists of the following components:

> * **MicroK8s** - for deploying local Kubernetes cluster that will host your OSM installation,

Reviewer: ```suggestion
* **MicroK8s** - for deploying a local Kubernetes cluster that will host your OSM installation,
```

> * **MicroStack** - for deploying local OpenStack cluster and using it as a VIM (Virtual Infrastructure Manager) for OSM,

Reviewer: ```suggestion
* **MicroStack** - for deploying a local OpenStack cluster and using it as a VIM (Virtual Infrastructure Manager) for OSM,
```

### In this tutorial you will learn how to:

* Deploy Charmed OSM on top of MicroK8s,
* Deploy MicroStack and add it as a VIM for Charmed OSM,
* Onboard sample VNF workloads through Charmed OSM.

### You will need:

A workstation which meets the following minimum recommended system requirements:
* **Ubuntu 18.04 LTS** Operating System,
* **16 GB** of RAM,
* **4 CPUs**,
* **50 GB** of free storage space.

## Before you start

Duration: 5:00

If you are using Windows or Mac OS, install [multipass][multipass] first to launch an Ubuntu VM.

If you are using Ubuntu 18.04 LTS, which we highly recommend, you can move on. Otherwise, you may need to install [snapd][snapd] first.

Install Juju and OSM client from the snap store:

```bash
$ sudo snap install juju --classic
juju 2.6.10 from Canonical✓ installed
$ sudo snap install osmclient --edge
osmclient (edge) 0+git.44be24a from Adam Israel (adam) installed
```

> Allows OSM client snap to access Juju configuration:

Reviewer: ```suggestion
Allow the OSM client snap to access the Juju configuration:
```

```bash
$ sudo snap connect osmclient:juju-client-observe
```

> Bootstrap Juju controller on LXD that OSM will use to deploy proxy charms:

Reviewer: ```suggestion
Bootstrap a Juju controller on LXD, that OSM will use to deploy proxy charms:
```

```bash
$ juju bootstrap localhost osm-lxd
Since Juju 2 is being run for the first time, downloading latest cloud information.
Fetching latest public cloud list...
Updated your list of public clouds with 2 cloud regions added:

    added cloud region:
        - google/asia-northeast2
        - google/europe-west6
Creating Juju controller "osm-lxd" on localhost/localhost
Looking for packaged Juju agent version 2.6.10 for amd64
To configure your system to better support LXD containers, please see: https://github.com/lxc/lxd/blob/master/doc/production-setup.md
Launching controller instance(s) on localhost/localhost...
 - juju-ba6acd-0 (arch=amd64)                 
Installing Juju agent on bootstrap instance
Fetching Juju GUI 2.15.0
Waiting for address
Attempting to connect to 10.25.201.34:22
Connected to 10.25.201.34
Running machine configuration script...
Bootstrap agent now started
Contacting Juju controller at 10.25.201.34 to verify accessibility...

Bootstrap complete, controller "osm-lxd" now is available
Controller machines are in the "controller" model
Initial model "default" added
```

positive
: This command takes a while to finish.

## Setting up MicroK8s

Duration: 3:00

> To install MicroK8s execute:

Reviewer: ```suggestion
To install MicroK8s, execute:
```

```bash
$ sudo snap install microk8s --classic
microk8s v1.16.2 from Canonical✓ installed
```

> Once completed, execute the following commands to set the required permissions and enable storage, and DNS plugins in MicroK8s:

Reviewer: ```suggestion
Once completed, execute the following commands to set the required permissions and enable the storage and DNS plugins in MicroK8s:
```

```bash
$ sudo usermod -a -G microk8s $USER
$ newgrp microk8s
$ microk8s.status --wait-ready

> microk8s is running

Reviewer: ```suggestion

microk8s is running
```

> addons:

Reviewer: We should probably crop this output, its content is not really relevant, eg.
```
microk8s is running
addons:
cilium: disabled
dashboard: disabled
...
Storage will be available soon
Enabling DNS
Applying manifest
serviceaccount/coredns created
configmap/coredns created
deployment.apps/coredns created
service/kube-dns created
clusterrole.rbac.authorization.k8s.io/coredns created
clusterrolebinding.rbac.authorization.k8s.io/coredns created
Restarting kubelet
DNS is enabled
```

serviceaccount/microk8s-hostpath created
clusterrole.rbac.authorization.k8s.io/microk8s-hostpath created
clusterrolebinding.rbac.authorization.k8s.io/microk8s-hostpath created
Storage will be available soon
Enabling DNS
Applying manifest
serviceaccount/coredns created
configmap/coredns created
deployment.apps/coredns created
service/kube-dns created
clusterrole.rbac.authorization.k8s.io/coredns created
clusterrolebinding.rbac.authorization.k8s.io/coredns created
Restarting kubelet
DNS is enabled
```

That’s it! Your local Kubernetes cluster is up and running. You can interact with it by using the `microk8s.kubectl` command, for example:

```bash
$ microk8s.kubectl cluster-info
Kubernetes master is running at https://127.0.0.1:16443
CoreDNS is running at https://127.0.0.1:16443/api/v1/namespaces/kube-system/services/kube-dns:dns/proxy

To further debug and diagnose cluster problems, use 'kubectl cluster-info dump'.
```

## Setting up MicroStack

Duration: 10:00

To install MicroK8s execute:

```bash
$ sudo snap install microstack --classic --beta
microstack (beta) stein from Canonical✓ installed
```

Once completed, execute the following commands:

```bash

> $ sudo microstack.init --auto

Reviewer: Same comment as above, can we crop the output in the middle?

  result = self._query(query)
Waiting for 10.20.20.1:8774
2019-11-10 20:56:26,974 - microstack_init - INFO - Creating default flavors...
2019-11-10 20:56:52,934 - microstack_init - INFO - Configuring Neutron
Waiting for 10.20.20.1:9696
2019-11-10 20:58:15,223 - microstack_init - INFO - Configuring Glance ...
Waiting for 10.20.20.1:9292
2019-11-10 20:58:50,777 - microstack_init - INFO - Fetching cirros image ...
100% [........................................................................] 12716032 / 127160322019-11-10 20:58:57,486 - microstack_init - INFO - Creating microstack keypair (~/.ssh/id_microstack)
2019-11-10 20:59:00,262 - microstack_init - INFO - Creating security group rules ...
2019-11-10 20:59:10,779 - microstack_init - INFO - restarting libvirt and virtlogd ...
2019-11-10 20:59:10,928 - microstack_init - INFO - Complete. Marked microstack as initialized!
```

That’s it! Your local OpenStack cluster is up and running. You can interact with it by using the `microstack.openstack` command.

> In order to configure MicroStack, so that it could be used as a VIM for Charmed OSM, add Ubuntu image:

Reviewer: ```suggestion
In order to configure MicroStack, so that it could be used as a VIM for Charmed OSM, you need to add an Ubuntu image:

Download Ubuntu Server 16.04 LTS with the following command:
```

```bash
$ wget https://cloud-images.ubuntu.com/releases/16.04/release/ubuntu-16.04-server-cloudimg-amd64-disk1.img
--2019-11-12 17:25:01--  https://cloud-images.ubuntu.com/releases/16.04/release/ubuntu-16.04-server-cloudimg-amd64-disk1.img
Resolving cloud-images.ubuntu.com (cloud-images.ubuntu.com)... 91.189.88.89, 2001:67c:1560:8001::8001
Connecting to cloud-images.ubuntu.com (cloud-images.ubuntu.com)|91.189.88.89|:443... connected.
HTTP request sent, awaiting response... 302 Found
Location: https://cloud-images.ubuntu.com/releases/xenial/release/ubuntu-16.04-server-cloudimg-amd64-disk1.img [following]
--2019-11-12 17:25:02--  https://cloud-images.ubuntu.com/releases/xenial/release/ubuntu-16.04-server-cloudimg-amd64-disk1.img
Reusing existing connection to cloud-images.ubuntu.com:443.
HTTP request sent, awaiting response... 200 OK
Length: 296878080 (283M) [application/octet-stream]
Saving to: ‘ubuntu-16.04-server-cloudimg-amd64-disk1.img.1’

ubuntu-16.04-server-cloudimg-amd64-disk1.img.1     100%[================================================================================================================>] 283.12M  24.2MB/s    in 10s     

> 2019-11-12 17:25:12 (27.7 MB/s) - ‘ubuntu-16.04-server-cloudimg-amd64-disk1.img.1’ saved [296878080/296878080]

Reviewer: Not needed.

> $ microstack.openstack image create \

Reviewer: ```suggestion
Then, pass it to the OpenStack cluster: 

```bash
$ microstack.openstack image create \
```

```

## Setting up Charmed OSM

Duration: 18:00

In this step we'll first deploy Charmed OSM on MicroK8s and then use MicroStack as a VIM.

### Deploy Charmed OSM

Bootstrap another Juju controller on MicroK8s that will host your Charmed OSM installation:

```bash
$ juju bootstrap microk8s osm-k8s
Creating Juju controller "osm-k8s" on microk8s/localhost
Creating k8s resources for controller "controller-osm-k8s"
Downloading images
Starting controller pod
Bootstrap agent now started
Contacting Juju controller at 10.152.183.19 to verify accessibility...

Bootstrap complete, controller "osm-k8s" now is available in namespace "controller-osm-k8s"

Now you can run
	juju add-model <model-name>
to create a new model to deploy k8s workloads.
```

positive
: This command takes a while to finish.

Once finished, add a model for OSM deployments:

```bash
$ juju add-model osm
Added 'osm' model on microk8s/localhost with credential 'microk8s' for user 'admin'
```

Generate an overlay bundle:

```bash
osmclient.overlay
```

Deploy Charmed OSM:

```bash
$ juju deploy osm --overlay vca-overlay.yaml

> Located bundle "cs:bundle/osm-21"

Reviewer: Can we crop?

10.152.183.236
```

### Add MicroStack as a VIM

In order to add MicroStack as a VIM for Charmed OSM, execute the following command:
```bash
$ osmclient.osm vim-create --name microstack \
      --user admin \
      --password keystone \
      --auth_url http://10.20.20.1:5000/v3 \
      --tenant admin \
      --account_type openstack \
      --config='{security_groups: default,
                 keypair: microstack,
                 project_name: admin,
                 user_domain_name: default,
                 region_name: microstack,
                 insecure: True,
                 availability_zone: nova,
                 version: 3,
                 use_floating_ip: true}'
c59b6823-65e5-4c64-9a36-86cea30ba1db
```

That’s it! You can start onboarding your VNFs.

## Onboarding VNFs

Duration: 15:00

At this point you have MicroK8s installed, Charmed OSM deployed on top of MicroK8s and MicroStack configured as a VIM. We will now deploy a basic network service to exercise the environment.

### Network Service

Download sample NSD (Network Service Descriptor) and VNFD (Virtual Network Function Descriptor):

```bash
$ wget http://bit.ly/basic_vnfd -O hackfest_basic_vnf.tar.gz

> --2019-11-12 16:48:16--  http://bit.ly/basic_vnfd

Reviewer: Output not needed, you should just put the two wget commands one after the other.

$ wget http://bit.ly/basic_nsd -O hackfest_basic_ns.tar.gz
--2019-11-12 16:49:00--  http://bit.ly/basic_nsd
Resolving bit.ly (bit.ly)... 67.199.248.10, 67.199.248.11
Connecting to bit.ly (bit.ly)|67.199.248.10|:80... connected.
HTTP request sent, awaiting response... 301 Moved Permanently
Location: http://osm-download.etsi.org/ftp/osm-5.0-five/6th-hackfest/packages/hackfest_basic_ns.tar.gz [following]
--2019-11-12 16:49:01--  http://osm-download.etsi.org/ftp/osm-5.0-five/6th-hackfest/packages/hackfest_basic_ns.tar.gz
Resolving osm-download.etsi.org (osm-download.etsi.org)... 195.238.226.47
Connecting to osm-download.etsi.org (osm-download.etsi.org)|195.238.226.47|:80... connected.
HTTP request sent, awaiting response... 200 OK
Length: 18463 (18K) [application/x-gzip]
Saving to: ‘hackfest_basic_ns.tar.gz’

hackfest_basic_ns.tar.gz                           100%[================================================================================================================>]  18.03K  --.-KB/s    in 0.1s    

2019-11-12 16:49:01 (189 KB/s) - ‘hackfest_basic_ns.tar.gz’ saved [18463/18463]

```

Then, upload the downloaded packages to OSM:

```bash

> $ osmclient.osm upload-package hackfest_basic_vnf.tar.gz

Reviewer: Same as above

d913b8e4-2f92-4c3d-bba0-2a3ad567190d
$ osmclient.osm upload-package hackfest_basic_ns.tar.gz
/snap/osmclient/37/lib/python2.7/site-packages/osmclient/common/utils.py:69: YAMLLoadWarning: calling yaml.load() without Loader=... is deprecated, as the default Loader is unsafe. Please read https://msg.pyyaml.org/load for full details.
  dict = yaml.load(tar.extractfile(yamlfile))
1eac876d-4406-4a55-b99a-baee9a672eb6
```

You can list the existing NSDs and VNFDs with the following commands:

```bash
$ osmclient.osm nsd-list
+-------------------+--------------------------------------+
| nsd name          | id                                   |
+-------------------+--------------------------------------+
| hackfest_basic-ns | 1eac876d-4406-4a55-b99a-baee9a672eb6 |
+-------------------+--------------------------------------+
$ osmclient.osm vnfd-list
+--------------------+--------------------------------------+
| nfpkg name         | id                                   |
+--------------------+--------------------------------------+
| hackfest_basic-vnf | d913b8e4-2f92-4c3d-bba0-2a3ad567190d |
+--------------------+--------------------------------------+
```

> Now, as the packages have been uploaded, you can create basic Network Service:

Reviewer: ```suggestion
Now, as the packages have been uploaded, you can create a basic Network Service:
```

Welcome to Ubuntu 16.04.6 LTS (GNU/Linux 4.4.0-166-generic x86_64)

 * Documentation:  https://help.ubuntu.com
 * Management:     https://landscape.canonical.com
 * Support:        https://ubuntu.com/advantage

0 packages can be updated.
0 updates are security updates.

The programs included with the Ubuntu system are free software;
the exact distribution terms for each program are described in the
individual files in /usr/share/doc/*/copyright.

Ubuntu comes with ABSOLUTELY NO WARRANTY, to the extent permitted by
applicable law.

To run a command as administrator (user "root"), use "sudo <command>".
See "man sudo_root" for details.

ubuntu@hackfest-basic-ns-1-hackfest-basic-vm-1:~$ 
```

### Network Service with Day-1/Day-2 configuration

The previous Network Service that we deployed was very basic. Now, we will introduce more complexity, by adding Day-1 and Day-2 configuration to the VNF being onboarded using Proxy Charms.

First, let's remove the existing Network Service:

```bash
$ osmclient.osm ns-delete acad055e-37f4-4b59-bb31-9a6c3b972be7
Deletion in progress
```

Then, download NS and VNF descriptors:

```bash

> $ wget http://bit.ly/simplecharm_vnfd -O hackfest_simplecharm_vnf.tar.gz

Reviewer: Same here, no need for the wget outputs

```

And check whether the file has been created:

```bash
$ IP=`microstack.openstack server list -c Networks -f value --name simplecharm | awk '{ print $2 }'`
$ ssh ubuntu@$IP -i .ssh/id_microstack ls
The authenticity of host '10.20.20.24 (10.20.20.24)' can't be established.
ECDSA key fingerprint is SHA256:PhIBx7jV8gTuGK5Hu/tqpEPdiTrg7zzGwHFa2CEbLT0.
Are you sure you want to continue connecting (yes/no)? yes
Warning: Permanently added '10.20.20.24' (ECDSA) to the list of known hosts.
first-touch
touched
```

## Enable/Disable the Development Stack

Duration: 2:00

We encourage you to disable the Development Stack when you are not using it, especially if you are running it on your workstation. This way, you do not waste your workstation’s resources. You can disable the Development Stack by running the following commands:

```bash
$ microk8s.disable dns
Disabling DNS
Reconfiguring kubelet
Removing DNS manifest
serviceaccount "coredns" deleted
configmap "coredns" deleted
deployment.apps "coredns" deleted
service "kube-dns" deleted
clusterrole.rbac.authorization.k8s.io "coredns" deleted
clusterrolebinding.rbac.authorization.k8s.io "coredns" deleted

> DNS is disabled

Reviewer: Not needed

$ sudo snap disable microstack

> microstack disabled

Reviewer: Not needed

$ sudo snap disable microk8s

> microk8s disabled

Reviewer: Not needed

```

You can always enable the Development Stack with the following commands:

```bash
$ sudo snap enable microstack; sleep 60
microstack enabled
$ sudo snap enable microk8s
microk8s enabled
$ microk8s.status --wait-ready
microk8s is running
addons:
cilium: disabled
dashboard: disabled
dns: enabled
fluentd: disabled
gpu: disabled
helm: disabled
ingress: disabled
istio: disabled
jaeger: disabled
knative: disabled
linkerd: disabled
metrics-server: disabled
prometheus: enabled
rbac: disabled
registry: disabled
storage: enabled
$ microk8s.enable dns
Enabling DNS
Applying manifest
serviceaccount/coredns created
configmap/coredns created
deployment.apps/coredns created
service/kube-dns created
clusterrole.rbac.authorization.k8s.io/coredns created
clusterrolebinding.rbac.authorization.k8s.io/coredns created
Restarting kubelet

> DNS is enabled

Reviewer: Not needed
